ably.ts.d

```diff
@@ -0,0 +1,457 @@
+    errorReason: ErrorInfo;
+    state: ChannelState;
+    presence: RealtimePresence;
+
+    attach: () => any;
+
+    detach:() => any;
+
+    history: (start: number,
+            end: number,
+            direction: String,
+            limit: number,
+            untilAttach: boolean) => PaginatedResult;
+
+    subscribe: (Msg: String | Message | MessageFn, ExtraMsg?: Message) => any;
+
+    unsubscribe: (Msg?: String | Message | MessageFn, ExtraMsg?: Message) => any;
+
+    publish: (name?: Array<String> | String, data?: any, clientId?: String, extras?: any) => any;
+
+  }
+
+  export declare class Channels {
+    exists: (name: String) => boolean;
+
+    get: (name: String, channelOptions?: ChannelOptions) => RealtimeChannel;
+    iterate: () => RealtimeChannel;
+
+    release: (name: String) => void;
+  }
+
+  export declare class Message {
+    constructor(name?: String, data?: any, clientId?: String);
+
+    fromEncoded: (JsonObject: String, channelOptions: ChannelOptions) => Message;
+
+    fromEncodedArray: (JsonArray: String, channelOptions: ChannelOptions) => Message;
+
+    clientId: String;
+    connectionId: String;
+    data: any;
+    encoding: String;
+    extras: String;
+    id: String;
+    name: String;
+    timestamp: number; // Time
+  }
+
+  export declare class PresenceMessage {
+    fromEncoded: (JsonObject: any, channelOptions?: ChannelOptions) => PresenceMessage;
+
+    fromEncodedArray: (JsonArray: Array<any>, channelOptions?: ChannelOptions) => Array<PresenceMessage>;
+
+    action: PresenceAction;
+    clientId: String;
+    connectionId: String;
+    data: any;
+    encoding: String;
+    id: String;
+    timestamp: number; // Time
+    memberKey: () => String;
+  }
+
+  export declare class ProtocolMessage {
+    action: ProtocolMessageAction;
+    auth: AuthDetails;
+    channel: String;
+    channelSerial: String;
+    connectionDetails: ConnectionDetails;
+    connectionId: String;
+    connectionKey: String;
+    connectionSerial: number;
+    count: number;
+    error: ErrorInfo;
+    flags: String;
+    id: String;
+    messages: Array<Message>;
+    msgSerial: number;
+    presence: Array<PresenceMessage>
+    timestamp: number; // Time
+  }
+
+  export declare class Rest {
+    constructor(options: String | ClientOptions);
+    auth: Auth;
+    channels: Channels;
+    request: (method: string, path: string, parmas?: any, body?: Array<any> | any, headers?: any) => HttpPaginatedResponse;
+    stats: ( start: number, // Time
+           end: number, // Time
+           direction: String,
+           limit: number,
+           unit: String) => PaginatedResult;
+    time: () => number; // time
+  }
+
+  export declare class Realtime {
+    constructor(options: String | ClientOptions);
+
+    auth: Auth;
+    channels: Channels;
+    clientId: String;
+    connection: Connection;
+
+    request: (method: string, path: string, parmas?: any, body?: Array<any> | any, headers?: any) => HttpPaginatedResponse;
```

Comment:
Out of interest, I think the callback def is actually `request => (err, response)` so is `err` not needed as it's "normal" for JS callbacks?